### PR TITLE
Improve write performance of the report generation within Tilkynna

### PR DESCRIPTION
**Pull request by melissapalmer**
**This PR includes various changes which will improve the report generation performance process within Tilkynna namely**
1) reuse destination spring application contexts
- which requires change to DB for destination to include __updated_on timestamp. So that this can be used to determine if a reload to get the new config is needed since last using the specified destination 
- addition of powermock as that mocking of static functions is possible

2) Restructure the way the reports where being pushed to threadpool for actually generating the reports. Logic is now as follows: 
- Scheduler is setup with 4 threads to poll DB for PENDING requests in batches of 30  (GenerateReportJobsAcquirer & GeneratedReportEntityRepository function getBatchOfPendingGenerateReportJobs)
- These are then pushed to a controlled threadpool  "generateReportThreadPoolExecutor" which has the threads actually generating the report and writting it out to the destination

3) I have also included logback config file into project such that we can log to diffrent files. I needed this to be able to trace the timings of report steps

### CODE REVIEW

File: src/main/java/org/tilkynna/report/generate/processengine/config/SchedulerConfig.java [hunk 1]
```diff
@@ -19,19 +19,29 @@
  */
 @Configuration
 public class SchedulerConfig implements SchedulingConfigurer {
-    private static final int POOL_SIZE = 15;
 
+    /**
+     * The pool size.
+     */
+    private static final int POOL_SIZE = 4;
```
Comment (mvniekerk) [inline src/main/java/org/tilkynna/report/generate/processengine/config/SchedulerConfig.java]: Not a configurable setting?

File: src/main/java/org/tilkynna/report/generate/processengine/config/SchedulerConfig.java [hunk 2]
```diff
@@ -19,19 +19,29 @@
...
+     *            The task registrar.
+     */
     @Override
     public void configureTasks(ScheduledTaskRegistrar scheduledTaskRegistrar) {
         ThreadPoolTaskScheduler threadPoolTaskScheduler = new ThreadPoolTaskScheduler();
 
         threadPoolTaskScheduler.setPoolSize(POOL_SIZE);
         threadPoolTaskScheduler.setThreadPriority(Thread.MIN_PRIORITY);
-        threadPoolTaskScheduler.setThreadNamePrefix("Scheduler-");
+        threadPoolTaskScheduler.setThreadNamePrefix("scheduled-task-pool-");
```
Comment (mvniekerk) [inline src/main/java/org/tilkynna/report/generate/processengine/config/SchedulerConfig.java]: Thread name prefix change inconsistent with ReportRequestThreadPoolConfig.java line 27 - I don't particularly care about the naming, just wondering why the change

File: src/main/resources/application.yml [hunk 3]
```diff
@@ -1,21 +1,23 @@
...
     threading:
-      poolSize: 15
-      maxPoolSize: 300
-      queueCapacity: 1000
+      poolSize: 30
+      maxPoolSize: 50
+      queueCapacity: 100
     monitorPendingRequests: #config for monitoring PENDING report requests and publishing them to be processed 
-      fixedRateInMilliseconds: 1000
+      #1000ms = 1second
```
Comment (mvniekerk) [inline src/main/resources/application.yml]: Wondering why the comment...

File: src/test/java/org/tilkynna/report/destination/mockdata/SMTPDestinationMockDataGenerator.java [hunk 4]
```diff
@@ -24,13 +24,12 @@ private static final SMTPDestinationEntity createSMTPDestinationEntity(String de
...
-        destination.setTimeout(new Long(
-                56322));
+        destination.setTimeout(new Long(56322));
         destination.setSecurityProtocol("securityProtocol");
+        destination.setUpdatedBy(UUID.randomUUID());
 
         destination.setHost("SMTP://localhost:5432");
-        destination.setPort(new Long(
-                23).shortValue());
+        destination.setPort(new Long(23).shortValue());
```
Comment (mvniekerk) [inline src/test/java/org/tilkynna/report/destination/mockdata/SMTPDestinationMockDataGenerator.java]: new Short() or new Integer()? Why the use of Long?

File: src/main/java/org/tilkynna/report/destination/integration/DynamicSftpChannelResolver.java [hunk 5]
```diff
@@ -18,10 +21,28 @@
...
+
 //based on https://github.com/spring-projects/spring-integration-samples/blob/master/advanced/dynamic-ftp/src/main/java/org/springframework/integration/samples/ftp/DynamicFtpChannelResolver.java
 // and http://scottfrederick.cfapps.io/blog/2012/05/22/Custom-PropertySource-in-Spring-3.1---Part-1
 @Component
-public class DynamicFtpChannelResolver {
+public class DynamicSftpChannelResolver {
+
+    @Setter
+    @Getter
+    public class Channel {
```
Comment (mvniekerk) [inline src/main/java/org/tilkynna/report/destination/integration/DynamicSftpChannelResolver.java]: This class is not referencing its enclosing / parent class' methods or members, thus can be marked static.

File: src/main/java/org/tilkynna/report/destination/integration/DynamicSftpChannelResolver.java [hunk 6]
```diff
@@ -18,10 +21,28 @@
...
+        private MessageChannel messageChannel;
+        private ZonedDateTime updatedOn;
+
+        public Channel(MessageChannel messageChannel, ZonedDateTime updatedOn) {
```
Comment (mvniekerk) [inline src/main/java/org/tilkynna/report/destination/integration/DynamicSftpChannelResolver.java]: By annotating the class with @AllArgsConstructor() allows you to remove this constructor.

File: src/main/java/org/tilkynna/report/destination/integration/DynamicSftpChannelResolver.java [hunk 7]
```diff
@@ -18,10 +21,28 @@
...
+            this.messageChannel = messageChannel;
+            this.updatedOn = updatedOn;
+        }
+
+    }
+
+    private final LinkedHashMap<UUID, Channel> channels = new LinkedHashMap<>();
```
Comment (mvniekerk) [inline src/main/java/org/tilkynna/report/destination/integration/DynamicSftpChannelResolver.java]: Rather program towards the interface java.util.map than to the concrete implementation LinkedHashMap

Comment (mvniekerk) [reply src/main/java/org/tilkynna/report/destination/integration/DynamicSftpChannelResolver.java]: Also, this hashmap is used in a concurrent environment - ConcurrentHashMap is more suited

File: src/main/java/org/tilkynna/report/destination/integration/DynamicSftpChannelResolver.java [hunk 8]
```diff
@@ -30,13 +51,40 @@
...
     public MessageChannel resolve(SFTPConfigSettings destination) {
-        MessageChannel channel = null;
-        if (channel == null) {
-            channel = createNewDestinationChannel(destination);
+        Channel channel = createOrUpdateDestinationChannel(destination);
+        return channel.getMessageChannel();
+    }
+
+    @SuppressWarnings("resource")
+    private synchronized Channel createOrUpdateDestinationChannel(SFTPConfigSettings destination) {
```
Comment (mvniekerk) [inline src/main/java/org/tilkynna/report/destination/integration/DynamicSftpChannelResolver.java]: The double-null-check with a synchronized on the second check will be more performant as you'll not be blocking threads.
e.g.:
Channel channel = this.channels.get(...);
boolean needsUpdate = channel == null || destination.getUpdateOn().is()...;
if (needsUpdate) {
    needsUpdate = channel == null || destination...is()..;
    if (needsUpdate) {
        synchronized (this.channels) {
           ... do the actual update
        }
    } 
}

Only blocking threads when you actually need to really update. The use of ConcurrentHashMap() will give you the guarantee that you'll not be mid-flight in updating the key's value if you're using it in a multithreaded environment.

File: src/main/java/org/tilkynna/report/generate/processengine/GenerateReportHandler.java [hunk 9]
```diff
@@ -55,21 +55,18 @@
...
-    @Async("generateReportTaskExecutor")
+    @Async("generateReportThreadPoolExecutor")
     @Transactional(propagation = Propagation.REQUIRES_NEW)
-    public void generateReportAsync(GeneratedReportEntity reportRequest) {
-        log.info(String.format("Report picked up by GenerateReportQueueHandler correlationId [%s] on Thread [%s]", reportRequest.getCorrelationId(), Thread.currentThread().getName()));
-
-        GeneratedReportEntity generatedReportEntity = generatedReportEntityRepository.findById(reportRequest.getCorrelationId()) //
-                .orElseThrow(() -> new ResourceNotFoundExceptions.GeneratedReportEntity(reportRequest.getCorrelationId().toString()));
+    public void pushGenerateReportToThreadPoolForProcessing(UUID correlationId) {
+        log.info(String.format("Report picked up by GenerateReportQueueHandler correlationId [%s] on Thread [%s]", correlationId, Thread.currentThread().getName()));
```
Comment (mvniekerk) [inline src/main/java/org/tilkynna/report/generate/processengine/GenerateReportHandler.java]: Logs are expensive - not convinced that these needs to be on .info level

File: src/main/java/org/tilkynna/report/generate/processengine/GenerateReportJobsAcquirer.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+    @Autowired
+    private GenerateReportHandler generateReportHandler;
+
+    @Autowired
+    @Qualifier("generateReportThreadPoolExecutor")
+    private Executor generateReportExecutor;
+
+    public void getPendingJobsAndPushToGenerateReportThreadPool() {
+        log.info("getPendingJobsAndPushToGenerateReportThreadPool START: {}", Thread.currentThread().getName());
```
Comment (mvniekerk) [inline src/main/java/org/tilkynna/report/generate/processengine/GenerateReportJobsAcquirer.java]: Same comment about log levels - these "we are here" info printouts are expensive. Suggest that it is on debug or trace level, and only ramp it up at runtime using JSX if needed.

File: src/main/resources/application.yml (hunk 3 above)
Comment (melissapalmer) [reply src/main/resources/application.yml]: So I don't need to think about what the 100 means

File: src/test/java/org/tilkynna/report/destination/mockdata/SMTPDestinationMockDataGenerator.java (hunk 4 above)
Comment (melissapalmer) [reply src/test/java/org/tilkynna/report/destination/mockdata/SMTPDestinationMockDataGenerator.java]: Just a way in the unit test to get the Short. There is no 'new Short(23)' option

File: src/main/java/org/tilkynna/report/generate/processengine/GenerateReportHandler.java (hunk 9 above)
Comment (melissapalmer) [reply src/main/java/org/tilkynna/report/generate/processengine/GenerateReportHandler.java]: This is wanted for Tilkynna so that there is:  'Our logging for failed reports must be at a sufficiently granular level for debugging and resolution in production.'

File: src/main/java/org/tilkynna/report/generate/processengine/GenerateReportJobsAcquirer.java (hunk 10 above)
Comment (melissapalmer) [reply src/main/java/org/tilkynna/report/generate/processengine/GenerateReportJobsAcquirer.java]: This is wanted for Tilkynna so that there is:  'Our logging for failed reports must be at a sufficiently granular level for debugging and resolution in production.'

Comment (mvniekerk) [reply src/main/java/org/tilkynna/report/generate/processengine/GenerateReportJobsAcquirer.java]: Rather have it as debug, and mark the class as printable on the log4j.xml

Comment (melissapalmer) [reply src/main/java/org/tilkynna/report/generate/processengine/GenerateReportJobsAcquirer.java]: Yes - this was done so that it is possible to trace what happened in production after the fact if reports go missing - as this has happened before

**Comment by melissapalmer** [comment]
Thanks @VictorSzewczenko  for your comments. 

**Answers First:** 
4) yes there is a retry mechanism, which moves either FAILED or 'stuck' jobs (STARTED for longer the x) back to PENDING

**Questions:** 
a) on point 1 'we cannot deduce which job is being processed where' this can be added in both approaches. Question is why do we need to know this, what advantage will it give? 
b) on point 5 'how do we ensure 1 pod/scheduler does not suck up all the jobs in the database'. Is this not the same question with a WorkerThread approach - how would you ensure that the set of WorkerThreads on one pod does not suck up all the jobs from DB? 
c) on point 3 'pod restarts cause queued jobs to be lost' yes: but same for WorkerThreads, any WorkThread that's picked up a Job but not yet completed it will be lost if restart (mitigated with the process for re-queuing items see answer to 4 below) 

**Discussions points:**
_With regards to the scheduler component vs. a set of worker threads. Both these approaches are just the trigger mechanism to say: acquired more jobs from the jobs DB table. Where pro's/con's:_ 
- Scheduler: max tigger is 1ms
- WorkerThreads: trigger is only as fast as the loop (could this faster than 1ms) 
- Scheduler: the Spring platform is managing the threads for us under the hood
- WorkerThreads: we need to manage these threads correctly for spinning up, shutdown, interuptions
- Scheduler: the threadpool of workers can grow/shrink when idling
- WorkerThreads: will always be this x number threads 

_The queue I have used for the ThreadPool is 'small' double the size of max number threads. Meaning that:_ 
- most jobs picked from the DB are pushed onto WorkerThreads immediately 
- if all workers are busy a couple are queue up such that they will immediately be able to start when 1st worker finishes rather than: query db, get job but no one is able to execute this job yet
- the small queue allows us to push jobs onto worker threads in batches as apposed to 1 at a time
- we don't move anything more to STARTED if this queue is full, which allows other pods to pick up jobs

**Comment by melissapalmer** [comment]
> The log levels are the concern why I am not approving it (yet)

@mvniekerk this has been corrected with https://github.com/GrindrodBank/tilkynna/commit/71dcf6fa50ec9cd0bfdff8db9917efb4a768999f

File: src/main/java/org/tilkynna/report/generate/processengine/config/SchedulerConfig.java (hunk 1 above)
Comment (melissapalmer) [reply src/main/java/org/tilkynna/report/generate/processengine/config/SchedulerConfig.java]: Changed to be configurable with https://github.com/GrindrodBank/tilkynna/commit/0611eb37fd5c497289ad22512d7407ee39ae5956

File: src/main/java/org/tilkynna/report/destination/integration/DynamicSftpChannelResolver.java (hunk 6 above)
Comment (melissapalmer) [reply src/main/java/org/tilkynna/report/destination/integration/DynamicSftpChannelResolver.java]: Using this annotation with https://github.com/GrindrodBank/tilkynna/commit/a86bc51841c5684523c488b8c9c88e7e9b469319

File: src/main/java/org/tilkynna/report/destination/integration/DynamicSftpChannelResolver.java (hunk 5 above)
Comment (melissapalmer) [reply src/main/java/org/tilkynna/report/destination/integration/DynamicSftpChannelResolver.java]: Changed with https://github.com/GrindrodBank/tilkynna/commit/30cbf6e5ee96832cb0bd306fc28874e42c313e2d

File: src/main/java/org/tilkynna/report/destination/integration/DynamicSftpChannelResolver.java (hunk 7 above)
Comment (melissapalmer) [reply src/main/java/org/tilkynna/report/destination/integration/DynamicSftpChannelResolver.java]: Changed with https://github.com/GrindrodBank/tilkynna/commit/30cbf6e5ee96832cb0bd306fc28874e42c313e2d

File: src/main/java/org/tilkynna/report/destination/integration/DynamicSftpChannelResolver.java (hunk 8 above)
Comment (melissapalmer) [reply src/main/java/org/tilkynna/report/destination/integration/DynamicSftpChannelResolver.java]: Changed with https://github.com/GrindrodBank/tilkynna/commit/30cbf6e5ee96832cb0bd306fc28874e42c313e2d

File: src/test/java/org/tilkynna/report/destination/mockdata/SMTPDestinationMockDataGenerator.java (hunk 4 above)
Comment (gm7) [reply src/test/java/org/tilkynna/report/destination/mockdata/SMTPDestinationMockDataGenerator.java]: how about `destination.setPort((short)23);` ?

File: src/main/java/org/tilkynna/report/destination/integration/DynamicSftpChannelResolver.java [hunk 11]
```diff
@@ -30,13 +48,49 @@
...
+    @SuppressWarnings("resource")
+    private Channel createOrUpdateDestinationChannel(SFTPConfigSettings destination) {
+        // Check if the SFTP settings have been changed since context for channel was setup, if yes new need to reload it
+        Channel channel = this.channels.get(destination.getDestinationId());
+
+        // double-null-check with a synchronized on the second check will be more performant as you'll not be blocking threads.
+        boolean needsUpdate = (channel == null) || destination.getUpdatedOn().isAfter(channel.getUpdatedOn());
+        if (needsUpdate) {
+            synchronized (this.channels) { // Only blocking threads when you actually need to really update.
+                needsUpdate = (channel == null) || destination.getUpdatedOn().isAfter(channel.getUpdatedOn());
```
Comment (mvniekerk) [inline src/main/java/org/tilkynna/report/destination/integration/DynamicSftpChannelResolver.java]: channel needs to be re-assigned here - otherwise same result for the boolean value

File: src/main/java/org/tilkynna/report/generate/processengine/GenerateReportHandler.java [hunk 12]
```diff
@@ -55,70 +55,67 @@
...
-            writeReportToDestination(reportRequest, destinationEntity, generatedReport);
+            writeReportToDestination(generatedReportEntity, generatedReportEntity.getDestination(), generatedReport);
 
             generatedReportEntity.setReportStatus(ReportStatusEntity.FINISHED);
 
             // TODO call the callback URL
 
         } catch (IOException | ParseException | BirtException | ReportDatasourceExceptionException | TemplateHasInactiveDatasourcesException knownException) {
-            log.info(String.format("knownException Generating Report correlationId [%s] on Thread [%s]", generatedReportEntity.getCorrelationId(), Thread.currentThread().getName()));
+            log.error(String.format("knownException Generating Report correlationId [%s] on Thread [%s]", generatedReportEntity.getCorrelationId(), Thread.currentThread().getName()));
```
Comment (mvniekerk) [inline src/main/java/org/tilkynna/report/generate/processengine/GenerateReportHandler.java]: Why .error and not .info?

Comment (melissapalmer) [reply src/main/java/org/tilkynna/report/generate/processengine/GenerateReportHandler.java]: because its an error -- in the catch statement
